drgn_tools/task.py

```diff
@@ -185,6 +195,173 @@ def get_command(task: Object) -> str:
     return escape_ascii_string(task.comm.string_())
 
 
+def get_ppid(task: Object) -> int:
+    """
+    :returns: Parent PID of the task
+    """
+    return task.parent.pid.value_()
+
+
+def build_rss_global(prog: Program) -> None:
```

Comment:
Ok, I did a thorough read through of the RSS accounting code, and I think we're completely overdoing this.

---

The whole "task_rss" and "mm_rss" thing just exists because of a performance optimization in the kernel. If these counters were only maintained as counters in the `mm_struct`, then there could be serious contention updating these values for processes with many threads. `SPLIT_RSS_COUNTING` allows each `task_struct` to contain a separate set of counters. For each task, every 64 page faults, the `mm_struct` gets its counters updated via `sync_mm_rss()`. This dramatically reduces the contention on the `mm_struct` counters.

In reality, we do not care about knowing a specific task's task_rss value. We _only_ care about the counters once you've combined the ones from the mm_struct and all of its associated task_structs. There's no need to cache any intermediate values. All we want is to have a single cache that maps an mm_struct to its final, computed RSS values.

There may be some value in keeping the filepages, anonpages, and shmempages separate. After all, `/proc/pid/status` breaks down these values too. I could see other helper code wanting to know this at some point. So let's try to keep the numbers separate, where possible.

And finally, there's one issue here: `build_rss_global()` computes data for every task in the system, which would be a waste. For an example, in the future, we would like to support an invocation like `ps <PID>` or `ps COMM`, which would only return rows for a subset of tasks. Computing the RSS for every single task would be a waste for those invocations.

Taking these points into account, here's a simplified and cleaned up implementation, maybe we could start with something like this instead?

```python
class TaskRss(NamedTuple):
    """Represent's a task's resident set size. See task_rss()."""

    file: int
    anon: int
    shmem: int

    @property
    def total(self) -> int:
        return self.file + self.anon + self.shmem


def task_rss(task: Object, cache: Optional[Dict[int, TaskRss]]) -> TaskRss:
    """
    Return the task's resident set size (RSS) in pages

    The task's RSS is the number of pages which are currently resident in
    memory. The RSS values can be broken down into anonymous pages (not bound to
    any file), file pages (those associated with memory mapped files), and
    shared memory pages (those which aren't associated with on-disk files, but
    belonging to shared memory mappings). This function returns a tuple
    containing each category, but the common behavior is to use the "total"
    value which sums them up.

    :param task: ``struct task_struct *`` for which to compute RSS
    :param cache: if provided, we can use this to cache the mapping of
      "mm_struct" to RSS. This helps avoid re-computing the RSS value for
      processes with many threads, but note that it could result in out of date
      values on a live system.
    :returns: the file, anon, and shmem page values
    """
    mmptr = task.mm.value_()
    if cache and mmptr in cache:
        return cache[mmptr]

    # Kthreads have a NULL mm, simply skip them, returning 0.
    if not task.mm:
        return TaskRss(0, 0, 0)

    prog = task.prog_
    MM_FILEPAGES = prog.constant("MM_FILEPAGES").value_()
    MM_ANONPAGES = prog.constant("MM_ANONPAGES").value_()
    try:
        MM_SHMEMPAGES = prog.constant("MM_SHMEMPAGES").value_()
    except LookupError:
        MM_SHMEMPAGES = -1

    # Start with the counters from the mm_struct
    file = anon = shmem = 0
    file += task.mm.rss_stat.count[MM_FILEPAGES].counter.value_()
    anon += task.mm.rss_stat.count[MM_ANONPAGES].counter.value_()
    if MM_SHMEMPAGES >= 0:
        shmem += task.mm.rss_stat.count[MM_SHMEMPAGES].counter.value_()

    # Augment the values with the ones from the task's rss_stat.
    # The task rss_stat is what gets updated directly, every so often these
    # changes get sync'd to the mm_struct.
    file += task.rss_stat.count[MM_FILEPAGES].value_()
    anon += task.rss_stat.count[MM_ANONPAGES].value_()
    if MM_SHMEMPAGES >= 0:
        shmem += task.rss_stat.count[MM_SHMEMPAGES].value_()

    # Finally, augment the values with the ones from the rest of the thread
    # group.
    for gtask in list_for_each_entry(
        "struct task_struct",
        task.thread_group.address_of_(),
        "thread_group",
    ):
        file += gtask.rss_stat.count[MM_FILEPAGES].value_()
        anon += gtask.rss_stat.count[MM_ANONPAGES].value_()
        if MM_SHMEMPAGES >= 0:
            shmem += gtask.rss_stat.count[MM_SHMEMPAGES].value_()

    rss = TaskRss(file, anon, shmem)
    if cache is not None:
        cache[mmptr] = rss
    return rss
```

The benefit of this `task_rss()` function is that it can be just as easily used as a helper function for other code. It works great on live systems (so long as you don't keep using the same `cache`), but it is efficient for the case of the `ps` command too.

Reply:
Thanks for the detailed explanation! We have implemented this in latest code.